hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/repository/BaseEntityRepositoryACM.java

```diff
@@ -0,0 +1,350 @@
+                                }
+                            } else if (Optional.class.isAssignableFrom(method.getReturnType())) {
+                                return ((Optional<T>)result).map(t -> isOperationAllowed(AccessController.Operation.READ, t, accessController));
+                            } else if (repository.getDomainClass().isAssignableFrom(method.getReturnType())) {
+                                accessController.assertOperationAllowed(AccessController.Operation.READ, (T)result);
+                            }
+                            return result;
+                        } else if ("toString".equals(method.getName()) && method.getParameterCount() == 0) {
+                            return BaseEntityRepositoryACM.class.getSimpleName() +
+                                    "(repository: " + repository + ", accessController: " +  accessController + ")";
+                        } else {
+                            return method.invoke(repository, args);
+                        }
+                    } catch (final InvocationTargetException e) {
+                        throw e.getCause() == null ? e : e.getCause();
+                    }
+                });
+        LOGGER.info("Proxy created -> {}", acmProxy);
+        return acmProxy;
+    }
+
+    @Override
+    @NonNull
+    public Optional<T> findById(@NonNull final Long id) {
+        return findOne(byIdSpec(id));
+    }
+
+    @Override
+    @NonNull
+    public List<T> findAll() {
+        return findAll((Specification<T>) null);
+    }
+
+    @Override
+    @NonNull
+    public List<T> findAllById(@NonNull final Iterable<Long> ids) {
+        return findAll(byIdsSpec(ids));
+    }
+
+    @Override
+    public boolean existsById(@NonNull final Long id) {
+        return exists(byIdSpec(id));
+    }
+
+    @Override
+    public long count() {
+        return count(null);
+    }
+
+    @Override
+    public void delete(@NonNull final T entity) {
+        accessController.assertOperationAllowed(AccessController.Operation.DELETE, entity);
+        repository.delete(entity);
+    }
+
+    @Override
+    public void deleteById(@NonNull final Long id) {
+        if (!exists(AccessController.Operation.READ, byIdSpec(id))) {
+          throw new EntityNotFoundException(repository.getDomainClass(), id);
+        }
+        if (!exists(AccessController.Operation.DELETE, byIdSpec(id))) {
+            throw new InsufficientPermissionException();
+        }
+        repository.deleteById(id);
+    }
+
+    @Override
+    public void deleteAllById(@NonNull final Iterable<? extends Long> ids) {
+        final List<Long> idList = toList(ids);
+        if (count(AccessController.Operation.DELETE, byIdsSpec(idList)) != count(AccessController.Operation.READ, byIdsSpec(idList))) {
+            throw new InsufficientPermissionException("Has at least one id that is not allowed for deletion!");
```

Comment:
Should we compare the counting results against the amount of provided id's as well?

Reply:
Actually maybe the best approach would be to compare only with idList.size instead of readable count
At the moment this check provides check that all that could be read  are with delete rights
However there could be inaccessible that would be deleted. 
What do you think to compare to idList.size?

Reply:
Yes, let's change it to compare the sizes DELETABLE.size == idList.size. However, let's map the `Iterable<? extends Long> ids` into a set instead of a `List<Long>` to get rid of potential duplicates.